### ISSUE TITLE
model.fit() appears to stop prematurely without iterating through the entire dataset. 

### ISSUE THREAD

**Issue by viveksck** [keras-team/keras#14648]
I am trying to pretrain BERT on a dataset of 65M examples for 1 epoch without early stopping (patience = 1). However I notice that model.fit just stops and finishes prematurely without any error. It does not go through the entire dataset. 

```
model.compile(
        loss=custom_bce_loss,
        optimizer=tf.keras.optimizers.SGD(learning_rate=0.01)
        metrics=[prauc_metric],
    )


early_stop_callback = tf.keras.callbacks.EarlyStopping(monitor="val_loss",
                                                       patience=1,
                                                       min_delta=0.01)
tensorboard_callback = tf.keras.callbacks.TensorBoard(
                            log_dir=tensorboard_path(model_name, base_dir),
                            update_freq="batch",
                            profile_batch=2,
                            embeddings_freq=0)

ckpt_callback = tf.keras.callbacks.ModelCheckpoint(
                      checkpoint_path(model_name, base_dir, ckpt_name_suffix),
                      monitor="val_loss",
                      save_best_only=False,
                      save_weights_only=False,
                      mode="min",  # minimize loss
                      save_freq=int(100000/BATCH_SIZE),
                      verbose=1)

custom_callback = CustomCallback()

# In[20]:


history = model.fit(train_dset,
                    epochs=1,
                    batch_size=BATCH_SIZE,
                    validation_data=val_dset,
                    callbacks=[early_stop_callback, 
                               tensorboard_callback, 
                               ckpt_callback])
`
```
The size of the dataset is 65 million instances and batch size is 64. So I would expect 1 epoch to have 1015625 steps. However, when I fit the model and training progresses it just finishes at 135585th step and says training is done (no errors). 
initially I thought that this might be because I was using Adam and the learning rate became too small and so gradient became 0 and so training stopped. So I tried using SGD with a constant learning rate and I still see the same issue. 

135585/135585 [==============================] - 22867s 168ms/step - loss: 1.4894 - pr_auc: 0.0030 - val_loss: 0.7146 - val_pr_auc: 0.0000e+00 

Tensorflow version used: In [3]: tf.__version__
Out[3]: '2.4.1'
I am using tf.keras.  Other details: The dataset (TFdataset) is constructed by streaming in data from Big Query and I have verified that iterating through the batches of the dataset (in a seperate for loop) yields 1015625 batches. I  am training this model on a A100 GPU.

**Comment by jvishnuvardhan** [comment]
@viveksck Can you please share a standalone code to reproduce the issue? Also, can you try running the code with recent TF2.6 or `tf-nightly` and post entire output. Thanks!
